Add a performance test for the batch interface of PlatformIO

cmcantalupo opened the pull request:
- Measure performance of the batch interface and print metrics to files
- The script does not enforce a performance target

cmcantalupo committed:
Performance test for batch read

- Count and delay option for the sample loop
- Bash script to run test in different environments
- Python script to visualize results

Co-authored-by: Daniel Wilson <daniel1.wilson@intel.com>
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd commented on service/integration/test/test_batch_perf.py:
```diff
@@ -0,0 +1,27 @@
+                (count, time) = line.split(',')
+                time_samples[cc].append(float(time))
+
+    height = [sum(vv)/len(vv) for vv in time_samples.values()]
```
Some of the cases spend more time on the first sample than on the rest of the samples (e.g., due to obtaining a session). Can we add error bars and/or separate bars for first&rest iterations so that information isn't lost in the mean delay?

cmcantalupo committed:
Add environment variables to disable uring and msr safe at runtime

- GEOPM_DISABLE_IO_URING will avoid async kernel io.
- GEOPM_DISABLE_MSR_SAFE will avoid using the msr_safe driver.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix bug where msr_batch would be used even if msr safe was disabled at build time

Co-authored-by: Daniel Wilson <daniel1.wilson@intel.com>
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Add plotter for test_batch_perf.sh

- Plot geopm batch read times by privilege level, msr backend, and first/rest iterations
- Add multiple runs to sh script and support newer plotters

Signed-off-by: Daniel Wilson <daniel1.wilson@intel.com>

bgeltz commented on service/src/IOGroup.cpp:
```diff
@@ -5,7 +5,7 @@
 
 #include <unistd.h>
-
+#include <cstdlib>
```
```suggestion
```

bgeltz commented on service/src/IOGroup.cpp:
```diff
@@ -99,7 +99,8 @@ namespace geopm
             // --enable-rawmsr and msr-safe driver is not available.
-            bool use_msr_safe = (access("/dev/cpu/msr_batch", R_OK | W_OK) == 0);
+            bool use_msr_safe = (access("/dev/cpu/msr_batch", R_OK | W_OK) == 0 &&
+                                 std::getenv("GEOPM_DISABLE_MSR_SAFE") == nullptr);
```
From Helper.[ch]pp:
```suggestion
                                 geopm::get_env("GEOPM_DISABLE_MSR_SAFE").empty);
```
Aside from the LevelZero.cpp checking ```ZE_AFFINITY_MASK``` directly with ```getenv``` (which imo is another issue we should correct), there are  *no* direct usages of ```getenv``` anywhere in the service code.

cmcantalupo replied:
We really want to know if the environment variable is set or not, we don't care about its contents.  In this way, the geopm::get_env() is ambiguous.  Since we are not using the C string in this case, I think the direct use of std::getenv() is valid.  If a user simply called `export GEOPM_DISABLE_MSR_SAFE` the original code would disable msr-safe use, but the proposed code would not.

bgeltz replied:
SGTM, then this PR needs to include some documentation about the expected usage of ```GEOPM_DISABLE_MSR_SAFE```.  IIRC we don't have other env vars that are exported without a value set, but I may be mistaken.

bgeltz commented on service/src/IOUring.cpp:
```diff
@@ -34,7 +36,8 @@ namespace geopm
 #ifdef GEOPM_HAS_IO_URING
-        if (IOUringImp::is_supported()) {
+        if (IOUringImp::is_supported() &&
+            std::getenv("GEOPM_DISABLE_IO_URING") == nullptr) {
```
Same suggestion as in IOGroup.cpp here.

cmcantalupo replied:
Same argument applies that this is preferred to geopm::get_env().

bgeltz commented on service/integration/test/plot_batch_perf.py:
```diff
@@ -0,0 +1,52 @@
+g.set(ylim=(args.min_y, None), xlabel=None)
+g.set_axis_labels(y_var='Latency')
+for ax in g.axes.flat:
+    ax.set_yscale('log')
```
Confused why this is set to 'log' while ```log_scale``` is commented out above.

cmcantalupo replied:
The commented out use of log_scale above didn't work properly (the bars became just a line where the top of the bar should have been).  Changing the scale afterward as above does preserve the bars.

cmcantalupo committed:
Document performance disable environment variables

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup: do not export SCRIPT_DIR

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
